File: src/state/商城/魔改道具.ts
import Actor from "@/Actor";
import AutoMissileAttackController from "@/AutoMissileAttackController";
import PlatUtil from "@/PlatUtil";
import Cache from "@/Cache";
import ActorItemUtil from "@/ActorItemUtil";
import HeroUtil from "@/HeroUtil";
import TextTagUtil from "@/TextTagUtil";
import PlayerUtil from "@/PlayerUtil";
import UnitStateUtil from "@/UnitStateUtil";
import EffectUtil from "@/EffectUtil";
import DamageUtil from "@/DamageUtil";
import ArrayUtil from "@/ArrayUtil";
import {d_核心部件} from "../../xlsx/装备/核心部件";
import BaseUtil from "@/BaseUtil";
import ActorUtil from "@/ActorUtil";

export default class 魔改道具 {


    constructor() {

        //团结号主炮 升级后立刻获得下次升级所需的80%经验值
        se.onHeroLevelUp((e, solarTrigger) => {
            let trigUnitOwner = e.trigUnitOwner;

            solarTrigger.enable = false;
            BaseUtil.runLater(0.01, () => {
                let unit = PlayerUtil.getHero(trigUnitOwner);
                if (PlatUtil.hasMallItem(trigUnitOwner, "CZMG46")) {
                    if (ActorUtil.isUnitHasActor(unit, "团结号主炮")) {
                        let neededXP = DzGetUnitNeededXP(unit, GetUnitLevel(unit)) - GetHeroXP(unit);
                        AddHeroXP(unit, neededXP * 0.6, true)
                    }
                }
                //每次进化后，获得当前经验20%的基础经验。
                if (PlatUtil.hasMallItem(trigUnitOwner, "CZDZC")) {
                    let neededXP = DzGetUnitNeededXP(unit, GetUnitLevel(unit)) - GetHeroXP(unit);
                    AddHeroXP(unit, neededXP * 0.2, true)
                }
                solarTrigger.enable = true;
            });
        })


    }

    static 船只魔改道具奖励(index: number, unit: unit) {
        let mgShopKey = "CZMG" + index;
        if (!PlatUtil.hasMallItem(GetOwningPlayer(unit), mgShopKey)) {
            return
        }
        if (index == 28) {
            //转职为天使号时获得5个胶胶果实
            ActorItemUtil.addActorItemForUnit("胶胶果实", unit, 5)
        } else if (index == 32) {
            //转职位皇家宝藏号时获得5万金币
            PlayerUtil.addGoldState(GetOwningPlayer(unit), 50000);
            TextTagUtil.textSuccess("获得5万金币", unit)
        } else if (index == 33) {
            //转职为雷雨号时，获得20%当前智力。
            HeroUtil.addHeroPropertyP(unit, false, 0, 0, 0.2);
            TextTagUtil.textSuccess("获得20%智力", unit)
        } else if (index == 34) {
            //转职为狂风呼啸号时，获得20%当前敏捷。
            HeroUtil.addHeroPropertyP(unit, false, 0, 0.2, 0);
            TextTagUtil.textSuccess("获得20%敏捷", unit)
        } else if (index == 36) {
            //  转职为最强攻击号时，获得20%当前攻击。
            UnitStateUtil.setDamageBase(unit, UnitStateUtil.getDamageBase(unit) * 1.2);
            TextTagUtil.textSuccess("获得20%攻击", unit)
        } else if (index == 37) {
            //  转职为兽族兵营号时，获得20%当前力量。
            HeroUtil.addHeroPropertyP(unit, false, 0.2, 0, 0);
            TextTagUtil.textSuccess("获得20%力量", unit)
        } else if (index == 38) {
            //          转职为人族兵营号时，获得20 % 当前护甲。
            UnitStateUtil.setArmor(unit, UnitStateUtil.getArmor(unit) * 1.2);
            TextTagUtil.textSuccess("获得20%护甲", unit)
        } else if (index == 39) {
            //  转职为暗夜精灵号时，获得20 % 当前生命值。
            UnitStateUtil.addMaxLifeAndLife(unit, UnitStateUtil.getMaxLife(unit) * 0.2);
            TextTagUtil.textSuccess("获得20%生命值", unit)
        } else if (index == 43) {
            //  转职成黑龙时，攻击1200范围内所有敌人一次。
            let damage = UnitStateUtil.calculateStateFormula({attack: 10}, unit);
            DamageUtil.damageEnemyUnitsInRange(unit, 1200, damage)
            EffectUtil.addSpecialEffectAndDestroy("model\\基本技能\\瘟疫号基础技能.mdx",
                GetUnitX(unit), GetUnitY(unit), 1);
        } else if (index == 44) {
            //  转职成神之号时，无敌30秒。
            UnitStateUtil.addInvulnerableIfNot(unit, 30);
            TextTagUtil.textSuccess("无敌30秒", unit)
        } else if (index == 50) {
            // 萧炎号可以直接随机获得一个核心部件
            let randomElement = ArrayUtil.randomElement(d_核心部件);
            ActorItemUtil.addActorItemForUnit(randomElement.id, unit)
        }


    }

    static

    主炮商城道具加强(actor: Actor) {
        let 拥有魔改道具 = 魔改道具.主炮是否拥有魔改道具(actor);
        if (拥有魔改道具 == false) {
            return;
        }

        let ctrl: AutoMissileAttackController = actor.autoMissileAttackController
        if (ctrl == null) {
            return
        }
        let index = actor.actorType.index;
        if (index == 0) {
            if (actor.attribute == null) {
                actor.attribute = {};
            }
            actor.attribute.add_strength_on_kill = (actor.attribute.add_strength_on_kill || 0) + 1;
            actor.attribute.add_agility_on_kill = (actor.attribute.add_agility_on_kill || 0) + 1;
            actor.attribute.add_intelligence_on_kill = (actor.attribute.add_intelligence_on_kill || 0) + 1;
        } else if (index == 5) {
            //黑珍珠号主炮的范围+200
            ctrl.range += 200;
        } else if (index == 6) {
            //黑色幽灵号附带真实伤害+5%
            ctrl.damageStateFormula.increased = (ctrl.damageStateFormula.increased || 0) + 0.05;
        } else if (index == 7 || index == 8) {
            //桑尼号的每次弹射伤害+10%
            // 阳光桑尼号的每次弹射伤害+10%
            ctrl.damageStateFormula.increased = (ctrl.damageStateFormula.increased || 0) + 0.05;
        } else if (index == 15) {
            //海皇波塞冬主炮范围+300
            ctrl.range += 300;
        } else if (index == 16 || index == 17) {
            //深海之女移速变为max
            // 武装直升机移速变为max
            if (actor.attribute == null) {
                actor.attribute = {};
            }
            actor.attribute.move_speed = 522;
        } else if (index == 19) {
            //商船杀敌金币额外+20
            if (actor.attribute == null) {
                actor.attribute = {};
            }
            actor.attribute.add_gold_on_kill = (actor.attribute.add_gold_on_kill || 0) + 20;
        } else if (index == 20) {
            //缅电船杀敌经验额外+1
            if (actor.attribute == null) {
                actor.attribute = {};
            }
            actor.attribute.杀敌经验 = (actor.attribute.杀敌经验 || 0) + 1;
        } else if (index == 22 || index == 23) {
            //魔法船技能伤害+35%
            // 顽石号技能伤害+35%
            if (actor.attribute == null) {
                actor.attribute = {};
            }
            actor.attribute.magic_damage_increased = (actor.attribute.magic_damage_increased || 0) + 0.35;
        } else if (index == 24) {
            //巡洋舰初始护甲+1000
            if (actor.attribute == null) {
                actor.attribute = {};
            }
            actor.attribute.def = (actor.attribute.def || 0) + 1000;
        } else if (index == 25) {
            //猩红之女初始敏捷+3000
            if (actor.attribute == null) {
                actor.attribute = {};
            }
            actor.attribute.agility = (actor.attribute.agility || 0) + 3000;
        } else if (index == 26) {
            //光明号主炮目标+1
            ctrl.targetCount++;
        } else if (index == 27) {
            //猩红之女初始敏捷+3000
            if (actor.attribute == null) {
                actor.attribute = {};
            }
            actor.attribute.attack_p = (actor.attribute.attack_p || 0) + 0.2;
        } else if (index == 29) {
            //船长号射程+500
            ctrl.range += 500;
        } else if (index == 30) {
            //安妮女王复仇号的主炮附带杀敌+2全属性
            if (actor.attribute == null) {
                actor.attribute = {};
            }
            actor.attribute.add_strength_on_kill = (actor.attribute.add_strength_on_kill || 0) + 2;
            actor.attribute.add_agility_on_kill = (actor.attribute.add_agility_on_kill || 0) + 2;
            actor.attribute.add_intelligence_on_kill = (actor.attribute.add_intelligence_on_kill || 0) + 2;
        } else if (index == 31) {
            //冒险号护甲+500
            if (actor.attribute == null) {
                actor.attribute = {};
            }
            actor.attribute.def = (actor.attribute.def || 0) + 500;
        } else if (index == 35) {
            //刀山号主炮射速+50%
            ctrl.coolDown *= 0.75;
        } else if (index == 40) {
            //不死亡灵号射程+500
            ctrl.range += 500;
        } else if (index == 45) {
            //剑圣主炮伤害范围+250
            ctrl.range += 250;
        }
    }


    static cache = new Cache();

    static 主炮是否拥有魔改道具(actor: Actor): boolean {
        if (actor.actorType.class != "主炮" && actor.actorType.class != "技能A") {
            return false;
        } else {
            // return true;
        }
        return 魔改道具.cache.get(actor.actorTypeId + actor.unitOwnerId, () => {
            let index = actor.actorType.index;
            let mgShopKey = "CZMG" + index;
            return PlatUtil.hasMallItem(actor.unitOwner, mgShopKey);
        });
    }


}

File: scripts/state/商城/魔改道具.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__New = ____lualib.__TS__New
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["7"] = 3,["8"] = 3,["9"] = 4,["10"] = 4,["11"] = 5,["12"] = 5,["13"] = 6,["14"] = 6,["15"] = 7,["16"] = 7,["17"] = 8,["18"] = 8,["19"] = 9,["20"] = 9,["21"] = 10,["22"] = 10,["23"] = 11,["24"] = 11,["25"] = 12,["26"] = 12,["27"] = 13,["28"] = 13,["29"] = 14,["30"] = 14,["31"] = 15,["32"] = 15,["33"] = 17,["34"] = 17,["35"] = 17,["37"] = 23,["38"] = 24,["39"] = 26,["40"] = 27,["41"] = 27,["42"] = 27,["43"] = 28,["44"] = 29,["45"] = 30,["46"] = 31,["47"] = 31,["48"] = 31,["49"] = 31,["50"] = 32,["53"] = 36,["54"] = 37,["55"] = 37,["56"] = 37,["57"] = 37,["58"] = 38,["60"] = 40,["61"] = 27,["62"] = 27,["63"] = 23,["64"] = 20,["65"] = 47,["66"] = 48,["67"] = 49,["68"] = 49,["69"] = 49,["70"] = 49,["73"] = 52,["74"] = 54,["75"] = 55,["76"] = 57,["77"] = 57,["78"] = 57,["79"] = 57,["80"] = 58,["81"] = 59,["82"] = 61,["83"] = 61,["84"] = 61,["85"] = 61,["86"] = 61,["87"] = 61,["88"] = 61,["89"] = 62,["90"] = 63,["91"] = 65,["92"] = 65,["93"] = 65,["94"] = 65,["95"] = 65,["96"] = 65,["97"] = 65,["98"] = 66,["99"] = 67,["100"] = 69,["101"] = 69,["102"] = 69,["103"] = 69,["104"] = 70,["105"] = 71,["106"] = 73,["107"] = 73,["108"] = 73,["109"] = 73,["110"] = 73,["111"] = 73,["112"] = 73,["113"] = 74,["114"] = 75,["115"] = 77,["116"] = 77,["117"] = 77,["118"] = 77,["119"] = 78,["120"] = 79,["121"] = 81,["122"] = 81,["123"] = 81,["124"] = 81,["125"] = 82,["126"] = 83,["127"] = 85,["128"] = 86,["129"] = 87,["130"] = 87,["131"] = 87,["132"] = 87,["133"] = 87,["134"] = 87,["135"] = 89,["136"] = 91,["137"] = 92,["138"] = 93,["139"] = 95,["140"] = 96,["142"] = 47,["143"] = 102,["144"] = 105,["145"] = 106,["148"] = 110,["149"] = 111,["152"] = 114,["153"] = 115,["154"] = 116,["155"] = 117,["157"] = 119,["158"] = 120,["159"] = 121,["160"] = 122,["161"] = 124,["162"] = 125,["163"] = 127,["164"] = 128,["165"] = 131,["166"] = 132,["167"] = 134,["168"] = 135,["169"] = 138,["170"] = 139,["172"] = 141,["173"] = 142,["174"] = 144,["175"] = 145,["177"] = 147,["178"] = 148,["179"] = 150,["180"] = 151,["182"] = 153,["183"] = 154,["184"] = 157,["185"] = 158,["187"] = 160,["188"] = 161,["189"] = 163,["190"] = 164,["192"] = 166,["193"] = 167,["194"] = 169,["195"] = 170,["197"] = 172,["198"] = 173,["199"] = 175,["200"] = 176,["201"] = 178,["202"] = 179,["204"] = 181,["205"] = 182,["206"] = 184,["207"] = 185,["208"] = 187,["209"] = 188,["211"] = 190,["212"] = 191,["213"] = 192,["214"] = 193,["215"] = 195,["216"] = 196,["218"] = 198,["219"] = 199,["220"] = 201,["221"] = 202,["222"] = 204,["223"] = 205,["224"] = 207,["226"] = 102,["227"] = 214,["228"] = 215,["229"] = 216,["232"] = 220,["233"] = 220,["234"] = 220,["235"] = 221,["236"] = 222,["237"] = 223,["238"] = 220,["239"] = 220,["240"] = 214,["241"] = 212});
local ____exports = {}
local ____PlatUtil = require("solar.solar-common.util.game.PlatUtil")
local PlatUtil = ____PlatUtil.default
local ____Cache = require("solar.solar-common.tool.Cache")
local Cache = ____Cache.default
local ____ActorItemUtil = require("solar.solar-common.actor.util.ActorItemUtil")
local ActorItemUtil = ____ActorItemUtil.default
local ____HeroUtil = require("solar.solar-common.util.unit.HeroUtil")
local HeroUtil = ____HeroUtil.default
local ____TextTagUtil = require("solar.solar-common.util.text.TextTagUtil")
local TextTagUtil = ____TextTagUtil.default
local ____PlayerUtil = require("solar.solar-common.util.game.PlayerUtil")
local PlayerUtil = ____PlayerUtil.default
local ____UnitStateUtil = require("solar.solar-common.util.unit.UnitStateUtil")
local UnitStateUtil = ____UnitStateUtil.default
local ____EffectUtil = require("solar.solar-common.util.game.EffectUtil")
local EffectUtil = ____EffectUtil.default
local ____DamageUtil = require("solar.solar-common.util.system.DamageUtil")
local DamageUtil = ____DamageUtil.default
local ____ArrayUtil = require("solar.solar-common.util.lang.ArrayUtil")
local ArrayUtil = ____ArrayUtil.default
local _____6838_5FC3_90E8_4EF6 = require("xlsx.装备.核心部件")
local ____d__6838_5FC3_90E8_4EF6 = _____6838_5FC3_90E8_4EF6["d_核心部件"]
local ____BaseUtil = require("solar.solar-common.util.BaseUtil")
local BaseUtil = ____BaseUtil.default
local ____ActorUtil = require("solar.solar-common.actor.util.ActorUtil")
local ActorUtil = ____ActorUtil.default
____exports.default = __TS__Class()
local _____9B54_6539_9053_5177 = ____exports.default
_____9B54_6539_9053_5177.name = "魔改道具"
function _____9B54_6539_9053_5177.prototype.____constructor(self)
    se:onHeroLevelUp(function(e, solarTrigger)
        local trigUnitOwner = e.trigUnitOwner
        solarTrigger.enable = false
        BaseUtil.runLater(
            0.01,
            function()
                local unit = PlayerUtil:getHero(trigUnitOwner)
                if PlatUtil:hasMallItem(trigUnitOwner, "CZMG46") then
                    if ActorUtil:isUnitHasActor(unit, "团结号主炮") then
                        local neededXP = DzGetUnitNeededXP(
                            unit,
                            GetUnitLevel(unit)
                        ) - GetHeroXP(unit)
                        AddHeroXP(unit, neededXP * 0.6, true)
                    end
                end
                if PlatUtil:hasMallItem(trigUnitOwner, "CZDZC") then
                    local neededXP = DzGetUnitNeededXP(
                        unit,
                        GetUnitLevel(unit)
                    ) - GetHeroXP(unit)
                    AddHeroXP(unit, neededXP * 0.2, true)
                end
                solarTrigger.enable = true
            end
        )
    end)
end
_____9B54_6539_9053_5177["船只魔改道具奖励"] = function(self, index, unit)
    local mgShopKey = "CZMG" .. tostring(index)
    if not PlatUtil:hasMallItem(
        GetOwningPlayer(unit),
        mgShopKey
    ) then
        return
    end
    if index == 28 then
        ActorItemUtil:addActorItemForUnit("胶胶果实", unit, 5)
    elseif index == 32 then
        PlayerUtil:addGoldState(
            GetOwningPlayer(unit),
            50000
        )
        TextTagUtil.textSuccess("获得5万金币", unit)
    elseif index == 33 then
        HeroUtil:addHeroPropertyP(
            unit,
            false,
            0,
            0,
            0.2
        )
        TextTagUtil.textSuccess("获得20%智力", unit)
    elseif index == 34 then
        HeroUtil:addHeroPropertyP(
            unit,
            false,
            0,
            0.2,
            0
        )
        TextTagUtil.textSuccess("获得20%敏捷", unit)
    elseif index == 36 then
        UnitStateUtil:setDamageBase(
            unit,
            UnitStateUtil:getDamageBase(unit) * 1.2
        )
        TextTagUtil.textSuccess("获得20%攻击", unit)
    elseif index == 37 then
        HeroUtil:addHeroPropertyP(
            unit,
            false,
            0.2,
            0,
            0
        )
        TextTagUtil.textSuccess("获得20%力量", unit)
    elseif index == 38 then
        UnitStateUtil:setArmor(
            unit,
            UnitStateUtil:getArmor(unit) * 1.2
        )
        TextTagUtil.textSuccess("获得20%护甲", unit)
    elseif index == 39 then
        UnitStateUtil:addMaxLifeAndLife(
            unit,
            UnitStateUtil:getMaxLife(unit) * 0.2
        )
        TextTagUtil.textSuccess("获得20%生命值", unit)
    elseif index == 43 then
        local damage = UnitStateUtil:calculateStateFormula({attack = 10}, unit)
        DamageUtil:damageEnemyUnitsInRange(unit, 1200, damage)
        EffectUtil:addSpecialEffectAndDestroy(
            "model\\基本技能\\瘟疫号基础技能.mdx",
            GetUnitX(unit),
            GetUnitY(unit),
            1
        )
    elseif index == 44 then
        UnitStateUtil:addInvulnerableIfNot(unit, 30)
        TextTagUtil.textSuccess("无敌30秒", unit)
    elseif index == 50 then
        local randomElement = ArrayUtil:randomElement(____d__6838_5FC3_90E8_4EF6)
        ActorItemUtil:addActorItemForUnit(randomElement.id, unit)
    end
end
_____9B54_6539_9053_5177["主炮商城道具加强"] = function(self, actor)
    local _____62E5_6709_9B54_6539_9053_5177 = ____exports.default["主炮是否拥有魔改道具"](____exports.default, actor)
    if _____62E5_6709_9B54_6539_9053_5177 == false then
        return
    end
    local ctrl = actor.autoMissileAttackController
    if ctrl == nil then
        return
    end
    local index = actor.actorType.index
    if index == 0 then
        if actor.attribute == nil then
            actor.attribute = {}
        end
        actor.attribute.add_strength_on_kill = (actor.attribute.add_strength_on_kill or 0) + 1
        actor.attribute.add_agility_on_kill = (actor.attribute.add_agility_on_kill or 0) + 1
        actor.attribute.add_intelligence_on_kill = (actor.attribute.add_intelligence_on_kill or 0) + 1
    elseif index == 5 then
        ctrl.range = ctrl.range + 200
    elseif index == 6 then
        ctrl.damageStateFormula.increased = (ctrl.damageStateFormula.increased or 0) + 0.05
    elseif index == 7 or index == 8 then
        ctrl.damageStateFormula.increased = (ctrl.damageStateFormula.increased or 0) + 0.05
    elseif index == 15 then
        ctrl.range = ctrl.range + 300
    elseif index == 16 or index == 17 then
        if actor.attribute == nil then
            actor.attribute = {}
        end
        actor.attribute.move_speed = 522
    elseif index == 19 then
        if actor.attribute == nil then
            actor.attribute = {}
        end
        actor.attribute.add_gold_on_kill = (actor.attribute.add_gold_on_kill or 0) + 20
    elseif index == 20 then
        if actor.attribute == nil then
            actor.attribute = {}
        end
        actor.attribute["杀敌经验"] = (actor.attribute["杀敌经验"] or 0) + 1
    elseif index == 22 or index == 23 then
        if actor.attribute == nil then
            actor.attribute = {}
        end
        actor.attribute.magic_damage_increased = (actor.attribute.magic_damage_increased or 0) + 0.35
    elseif index == 24 then
        if actor.attribute == nil then
            actor.attribute = {}
        end
        actor.attribute.def = (actor.attribute.def or 0) + 1000
    elseif index == 25 then
        if actor.attribute == nil then
            actor.attribute = {}
        end
        actor.attribute.agility = (actor.attribute.agility or 0) + 3000
    elseif index == 26 then
        ctrl.targetCount = ctrl.targetCount + 1
    elseif index == 27 then
        if actor.attribute == nil then
            actor.attribute = {}
        end
        actor.attribute.attack_p = (actor.attribute.attack_p or 0) + 0.2
    elseif index == 29 then
        ctrl.range = ctrl.range + 500
    elseif index == 30 then
        if actor.attribute == nil then
            actor.attribute = {}
        end
        actor.attribute.add_strength_on_kill = (actor.attribute.add_strength_on_kill or 0) + 2
        actor.attribute.add_agility_on_kill = (actor.attribute.add_agility_on_kill or 0) + 2
        actor.attribute.add_intelligence_on_kill = (actor.attribute.add_intelligence_on_kill or 0) + 2
    elseif index == 31 then
        if actor.attribute == nil then
            actor.attribute = {}
        end
        actor.attribute.def = (actor.attribute.def or 0) + 500
    elseif index == 35 then
        ctrl.coolDown = ctrl.coolDown * 0.75
    elseif index == 40 then
        ctrl.range = ctrl.range + 500
    elseif index == 45 then
        ctrl.range = ctrl.range + 250
    end
end
_____9B54_6539_9053_5177["主炮是否拥有魔改道具"] = function(self, actor)
    if actor.actorType.class ~= "主炮" and actor.actorType.class ~= "技能A" then
        return false
    else
    end
    return ____exports.default.cache:get(
        actor.actorTypeId .. tostring(actor.unitOwnerId),
        function()
            local index = actor.actorType.index
            local mgShopKey = "CZMG" .. tostring(index)
            return PlatUtil:hasMallItem(actor.unitOwner, mgShopKey)
        end
    )
end
_____9B54_6539_9053_5177.cache = __TS__New(Cache)
return ____exports
